src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -206,5 +291,160 @@ public Task UnloadAsync()
+
+                // Nuget should wait until the project at least nominates once.
+                if (SourceBlock.Completion.IsCompleted)
+                {
+                    return false;
+                }
+
+                // After the first nomination, we should check the saved nominated version
+                return IsSavedNominationEmptyOrOlder(activeConfiguredProject);
+            }
+        }
+
+        private bool IsNuGetRestoreOnGoing()
+        {
+            // If NominateForRestoreAsync() has not finished, return false
+            return _nugetRestoreStarted;
+        }
+
+        protected virtual bool IsSavedNominationEmptyOrOlder(ConfiguredProject activeConfiguredProject)
+        {
+            if (!_savedNominatedConfiguredVersion.TryGetValue(activeConfiguredProject.ProjectConfiguration,
+                    out IComparable latestSavedVersionForActiveConfiguredProject) ||
+                activeConfiguredProject.ProjectVersion.IsLaterThan(latestSavedVersionForActiveConfiguredProject))
+            {
+                return true;
+            }
+
+            foreach (var loadedProject in activeConfiguredProject.UnconfiguredProject.LoadedConfiguredProjects)
+            {
+                if (_savedNominatedConfiguredVersion.TryGetValue(loadedProject.ProjectConfiguration, out IComparable savedProjectVersion))
+                {
+                    if (loadedProject.ProjectVersion.IsLaterThan(savedProjectVersion))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        protected virtual bool IsProjectConfigurationVersionOutOfDate(IReadOnlyCollection<PackageRestoreConfiguredInput>? configuredInputs)
+        {
+            Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
+            Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
+
+            if (configuredInputs is null)
+            {
+                return false;
+            }
+
+            var activeConfiguredProject = _project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+
+            IComparable? activeProjectConfigurationVersionFromConfiguredInputs = null;
+            foreach (var configuredInput in configuredInputs)
+            {
+                if (configuredInput.ProjectConfiguration.Equals(activeConfiguredProject.ProjectConfiguration))
+                {
+                    activeProjectConfigurationVersionFromConfiguredInputs = configuredInput.ConfiguredProjectVersion;
+                }
+            }
+
+            if (activeProjectConfigurationVersionFromConfiguredInputs is null ||
+                activeConfiguredProject.ProjectVersion.IsLaterThan(
+                    activeProjectConfigurationVersionFromConfiguredInputs))
+            {
+                return true;
+            }
+
```

Comment:
the function looks correct to me.  my suggestion is that we can add a quick check before line 435 (double foreach loop)

if (configuredInputs.Count == 1)
{
 return true;
}

The reason is if it contains a single configuration, it must be the active configuration compared earlier, so there is no need do any further check.  That will prevent us to run into the double foreach loop for majority of single target framework projects. 

It is not essential, but helpful for perf.